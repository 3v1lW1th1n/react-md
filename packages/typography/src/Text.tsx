import * as React from "react";
import * as PropTypes from "prop-types";
import cn from "classnames";

/**
 * A union of all the material design provided typography styles. When used with the Text
 * component, this will generate the correct typography className to apply and determine
 * what component to be rendered as if none was provided.
 */
export type TextTypes =
  | "headline-1"
  | "headline-2"
  | "headline-3"
  | "headline-4"
  | "headline-5"
  | "headline-6"
  | "subtitle-1"
  | "subtitle-2"
  | "body-1"
  | "body-2"
  | "caption"
  | "overline"
  | "button";

/**
 * A list of the default supported elements that the `Text` component can be rendered as. This is mostly used
 * for adding the correct HTMLAttributes and enabling the forward ref.
 */
export type DefaultTextElement =
  | HTMLHeadingElement
  | HTMLParagraphElement
  | HTMLSpanElement
  | HTMLDivElement
  | HTMLButtonElement
  | HTMLTableCaptionElement
  | HTMLAnchorElement
  | HTMLBodyElement
  | HTMLHtmlElement;

/**
 * The default additional props that can be applied to the Text component. This mostly just
 * covers all the elements that can be rendered "natively".
 */
export type DefaultTextProps = React.HTMLAttributes<DefaultTextElement>;

/**
 * A type describing the text component's children render function. It provides an object containing
 * the correct (and merged) className and expects a renderable element to be returned.
 */
export type TextRenderFunction = (props: { className: string }) => React.ReactNode;

/**
 * The base props for rendering the text component.
 *
 * @typeparam P - Any additional props that are available based on the component prop.
 */
export interface ITextProps<P = DefaultTextProps> {
  /**
   * An optional className to merge into typography styles.
   */
  className?: string;

  /**
   * The component to render as when the children are not a render function. If this prop is omitted,
   * the component will be determined by the `type` prop where:
   * - `"headline-1" -> <h1>`
   * - `"headline-2" -> <h2>`
   * - `"headline-3" -> <h3>`
   * - `"headline-4" -> <h4>`
   * - `"headline-5" -> <h5>`
   * - `"headline-6" -> <h6>`
   * - `"subtitle-1" -> <h5>`
   * - `"subtitle-2" -> <h6>`
   * - `"body-1"     -> <p>`
   * - `"body-2"     -> <p>`
   * - `"caption"    -> <caption>`
   * - `"overline"   -> <span>`
   * - `"button"     -> <button>`
   *
   */
  component?: React.ReactType<P> | null;

  /**
   * One of the material design typography text styles. This is used to generate a className
   * that can be applied to any element and have the correct typography.
   */
  type?: TextTypes;

  /**
   * Either a child render function or a react node. If this is not the child render function, a
   * different wrapper component can be provided using the `component` prop.
   */
  children?: React.ReactNode | TextRenderFunction;

  /**
   * @private
   */
  forwardedRef?: React.Ref<HTMLSpanElement>;
}

/**
 * The default defined props for the text component.
 */
export interface ITextDefaultProps {
  type: TextTypes;
  component: React.ReactType | null;
}

/**
 * @private
 */
export type TextWithDefaultProps<P = DefaultTextProps> = ITextProps<P> & ITextDefaultProps;

/**
 * The `Text` component is used to render text with the material design typography styles applied. By
 * default, everything will be rendered in a `<p>` tag with the normal paragraph styles.
 *
 * When the `type` prop is changed to another typography style, this component will determine the "best"
 * element to render the text in *unless* the `component` prop is provided. The default mapping is:
 * - `"headline-1" -> <h1>`
 * - `"headline-2" -> <h2>`
 * - `"headline-3" -> <h3>`
 * - `"headline-4" -> <h4>`
 * - `"headline-5" -> <h5>`
 * - `"headline-6" -> <h6>`
 * - `"subtitle-1" -> <h5>`
 * - `"subtitle-2" -> <h6>`
 * - `"body-1"     -> <p>`
 * - `"body-2"     -> <p>`
 * - `"caption"    -> <caption>`
 * - `"overline"   -> <span>`
 * - `"button"     -> <button>`
 * NOTE: if the `component` prop is not `null`, this logic will be ignored and the provided `component`
 * will be used instead.
 *
 * @forwardRef
 * @typeparam P - Any additional props that are valid when using the `component` prop or the built-in
 * "auto-component" logic. By default, this will just allow any HTMLElement props for each the default
 * elements in the "auto-component" logic.
 */
class Text<P extends {} = DefaultTextProps> extends React.Component<ITextProps<P> & P> {
  public static propTypes = {
    style: PropTypes.object,
    className: PropTypes.string,
    component: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
    type: PropTypes.oneOf([
      "headline-1",
      "headline-2",
      "headline-3",
      "headline-4",
      "headline-5",
      "headline-6",
      "subtitle-1",
      "subtitle-2",
      "body-1",
      "body-2",
      "caption",
      "overline",
      "button",
    ]),
    children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  };

  public static defaultProps: ITextDefaultProps = {
    type: "body-1",
    component: null,
  };

  public render() {
    const { className: propClassName, component, type, children, forwardedRef, ...props } = this
      .props as TextWithDefaultProps<DefaultTextProps>;

    const className = cn(`rmd-typography rmd-typography--${type}`, propClassName);
    if (typeof children === "function") {
      return (children as TextRenderFunction)({ className });
    }

    const Component = this.getComponent(component, type);
    return (
      <Component {...props} ref={forwardedRef} className={className}>
        {children}
      </Component>
    );
  }

  /**
   * A utility function to get the html tag the Text component should render as. This component will
   * attempt to render as the provided `component` or some auto-logic for determine what html tag
   * should be used for styling. All fallbacks will be to place the children in a span element.
   */
  private getComponent = (component: React.ReactType<P> | null, type: TextTypes) => {
    if (component) {
      return component;
    }

    switch (type) {
      case "headline-1":
        return "h1";
      case "headline-2":
        return "h2";
      case "headline-3":
        return "h3";
      case "headline-4":
        return "h4";
      case "headline-5":
        return "h5";
      case "headline-6":
      case "subtitle-1":
      case "subtitle-2":
        return "h6";
      case "body-1":
      case "body-2":
        return "p";
      case "caption":
        return "caption";
      case "button":
        return "button";
      default:
        return "span";
    }
  };
}

export default React.forwardRef<DefaultTextElement, ITextProps>((props, ref) => (
  <Text {...props} forwardedRef={ref} />
));
