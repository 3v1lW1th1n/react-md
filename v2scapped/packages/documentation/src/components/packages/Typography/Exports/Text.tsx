import * as React from "react";
import { Text, TextContainer } from "@react-md/typography";
import { uniq } from "lodash";

import { Markdown } from "components/Markdown";
import { DocumentationPage, Banner } from "components/DocumentationPage";
import { IDocumentedFile, IDocumentedComponent } from "components/Exports/types.d";

import "./text.scss";
import { textFile } from "./dummyData";
import Props from "./Props";

export interface ITextProps {
  file?: IDocumentedFile;
}

export interface ITextDefaultProps {
  file: IDocumentedFile;
}

export type TextWithDefaultProps = ITextProps & ITextDefaultProps;

export default class TextDocumentation extends React.Component<ITextProps> {
  public static defaultProps: ITextDefaultProps = {
    file: textFile,
  };

  public render() {
    const { file } = this.props as TextWithDefaultProps;
    const exports = Object.keys(file).map(key => file[key]);
    const component = exports.find(({ type }) => type === "component") as IDocumentedComponent;
    const { name, description, source, sourceLine } = component;
    return (
      <DocumentationPage>
        <Banner source={source} sourceLine={sourceLine}>
          {name}
        </Banner>
        <Markdown markdown={description} className="elevated" />
        <Props component={component} file={file} />
      </DocumentationPage>
    );
  }
}
