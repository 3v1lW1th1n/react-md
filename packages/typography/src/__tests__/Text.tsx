import * as React from "react";
import { create } from "react-test-renderer";

import Text, { TextTypes, TextRenderFunction } from "../Text";
import { mount } from "enzyme";

describe("Text", () => {
  it("should render correctly based on type prop", () => {
    const tests: { type: TextTypes; expected: string }[] = [
      { type: "headline-1", expected: "h1" },
      { type: "headline-2", expected: "h2" },
      { type: "headline-3", expected: "h3" },
      { type: "headline-4", expected: "h4" },
      { type: "headline-5", expected: "h5" },
      { type: "headline-6", expected: "h6" },
      { type: "subtitle-1", expected: "h6" },
      { type: "subtitle-2", expected: "h6" },
      { type: "body-1", expected: "p" },
      { type: "body-2", expected: "p" },
      { type: "overline", expected: "span" },
      { type: "button", expected: "button" },
    ];

    tests.forEach(({ type, expected }) => {
      const text = mount(<Text type={type} />);
      expect(text.find(expected).length).toBe(1);
      expect(text.find(expected).hasClass("rmd-typography")).toBe(true);
      expect(text.find(expected).hasClass(`rmd-typography--${type}`)).toBe(
        true
      );
    });

    const text = mount(
      <table>
        <Text type="caption" />
      </table>
    );
    expect(text.find("caption").length).toBe(1);
  });

  it("should default to rendering as a paragraph tag with the body-1 styles", () => {
    const text = mount(<Text />);

    expect(text.find("p").length).toBe(1);
    expect(text.find("p").hasClass("rmd-typography--body-1")).toBe(true);
  });

  it("should be able to render as a string component prop", () => {
    const component = "section";
    const text = mount(<Text component={component} />);
    expect(text.find(component).length).toBe(1);
  });

  it("should be able to render as a functional component", () => {
    const Component = ({ className, children }: any) => (
      <section className={`${className} custom`}>{children}</section>
    );

    const text = mount(<Text component={Component}>Hello, world!</Text>);
    expect(text.find(Component).length).toBe(1);
    expect(text.render()).toMatchSnapshot();
  });

  it("should be able to render as a class component", () => {
    class Component extends React.Component<any> {
      public render() {
        const { children, className } = this.props;
        return <section className={`${className} custom`}>{children}</section>;
      }
    }

    const text = mount(<Text component={Component}>Hello, world!</Text>);
    expect(text.find(Component).length).toBe(1);
    expect(text.render()).toMatchSnapshot();
  });

  it("should be able to use a children render function", () => {
    expect(
      create(
        <Text>
          {({ className }) => (
            <span className={`${className} span`}>Content</span>
          )}
        </Text>
      ).toJSON()
    ).toMatchSnapshot();
  });
});
