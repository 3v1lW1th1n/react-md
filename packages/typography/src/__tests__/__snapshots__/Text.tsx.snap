// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should be able to render as a class component 1`] = `
<section
  class="rmd-typography rmd-typography--body-1 custom"
>
  Hello, world!
</section>
`;

exports[`Text should be able to render as a functional component 1`] = `
<section
  class="rmd-typography rmd-typography--body-1 custom"
>
  Hello, world!
</section>
`;

exports[`Text should be able to use a children render function 1`] = `
<span
  className="rmd-typography rmd-typography--body-1 span"
>
  Content
</span>
`;
